File: src/main/java/org/frc1778/freezylib/logging/Field.java
package org.frc1778.freezylib.logging;

import com.google.gson.annotations.Expose;
import org.frc1778.freezylib.util.Measurement;

public abstract class Field {

  @Expose private String name;
  @Expose private String unit;
  @Expose private String type;

  public void setName(String name) {
    this.name = name;
  }

  public void setUnit(String unit) {
    this.unit = unit;
  }

  public void setUnit(Measurement.Base unit) {
    this.unit = unit.getSymbol();
  }

  public void setType(Class<?> type) {
    this.type = type.getName();
  }

  public String getName() {
    return name;
  }

  public String getUnit() {
    return unit;
  }

  public String getType() {
    return type;
  }

  public abstract String getValue();
}

File: src/main/kotlin/org/frc1778/freezylib/logging/Field.kt
package org.frc1778.freezylib.logging

import com.google.gson.annotations.Expose
import kotlin.reflect.KClass
import org.frc1778.freezylib.util.Measurement

abstract class Field {

    @Expose var name = ""
    @Expose var symbol = ""
    @Expose var unit: Measurement.Base = Measurement.Unitless.UNITLESS
        set(value) {
        field = value
        this.symbol = value.symbol
    }
    @Expose lateinit var type: String

    abstract var value: String

    fun setType(type: KClass<*>) {
        this.type = type.toString()
    }
}
